Use k8s keychain to auth against IaaS

DennisDenuto opened the pull request.

DennisDenuto committed:
feat: add k8s keychain

- k8s_keychain retrieves registry auth from kubernetes secrets as well
as registers IaaS providers to retrieve IaaS auth from its corresponding
metadata service

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
test refactor:

- use fake registry to avoid issues if the host has a dockerconfig file
- update test names

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/imgpkg/registry/keychain.go:
```diff
@@ -20,7 +22,11 @@ type KeychainOpts struct {
+	if err != nil {
+		panic(err.Error())
+	}
+	return regauthn.NewMultiKeychain(k8sKeychain, &envKeychain{environFunc: environFunc}, customRegistryKeychain{opts: keychainOpts})
```
something to think about:
- where should k8s keychain be inserted. should credentials provided explicitly take precedence over implicit creds?
- it seems that k8s keychain does quite a bit extra -- it does k8s specific lookups in addition to iaas provided creds. i dont think we want to include k8s lookups.
- ive been clicking around and stumbled upon https://github.com/kubernetes/kubernetes/issues/86245 along side https://github.com/vdemeester/k8s-pkg-credentialprovider/issues.

DennisDenuto replied:
> should credentials provided explicitly take precedence over implicit creds

The order is going to be env -> iaas -> cli -> config.json

iaas takes precedence over cli because iaas is targeted to a registry hostname (while cli isn't)

> k8s keychain does quite a bit extra

True, I removed the k8s keychain and instead will register the iaas providers via *magic imports*. Also this reduces the number of vendor files. And makes testing a bit simpler too.

> ive been clicking around and stumbled upon ...

good catch. I have added logic to avoid waiting forever during initializing the iaas keychain :-) 

Also, I added retry logic when getting the `Authorization` credentials from the metadata service too.

DennisDenuto committed:
feat: Add ordering of iaas

- env > iaas > cli > config.json

Authored-by: Dennis Leon <leonde@vmware.com>

Test refactor

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
refactor:

- Move auth keychains to registry/auth package
- add godoc for functions
- use a channel to block provider for testing

Authored-by: Dennis Leon <leonde@vmware.com>

inline iaas keychain

- removes k8s auth logic
- removes k8s library dep

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
feat: add feature flag

- IMGPKG_ENABLE_IAAS_AUTH receives a boolean, and toggles the iaas auth
on/off

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
refactor:

- move timeout logic into iaas keychain constructor

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX: improve iaas auth related errors

- provide error if feature flag for iaas is not a valid bool value
- return an error if timeout occurs trying to enable IaaS metadata service

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
test refactor:

- speed up tests by not waiting 15 seconds for the iaas keychain to timeout

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX: improve iaas auth related errors

- improve error message when a non-bool value is given to the iaas auth feature flag
- registry will return an error if unable to create a keychain instead of panic

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
feat: Add aws and azure iaas auth providers

- magic import includes init() functions that register the provider.
These registered providers are 'picked up' by the iaas_keychain logic
used by imgpkg's registry

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
feat: add retry logic fetching iaas auth creds

- This is just in case the IaaS metadata is flakey when getting the docker config for an image.
-- The iaas_keychain has already found credentials for the image. So if
during the 'Authorization' function we cannot find the creds for the
image, we can assume it is due to flakiness and retry up to 5 times

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto commented:
TODO:

- [x] fix up commits /squash
- [x] manual test imgpkg in Azure vm
- [x] manual test imgpkg in GCP vm
- [x] `--registry-verify-certs=false` on gcp. why? we want to avoid this
- [x] Handle logs being printed by iaas 
- [x] add documentation for new `IMGPKG_ENABLE_IAAS_AUTH` feature flag and the azure flag
- [ ] manual test imgpkg in AWS vm
- [ ] make the UX around the azure flag more compatible with kapp-controller (work with @cppforlife on this)

Logs from providers have been suppressed:

i.e. BEFORE:
```
E0908 18:01:44.291155    3447 aws_credentials.go:77] while getting AWS credentials NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
E0908 18:01:44.291315    3447 azure_credentials.go:201] Failed to create service principal token: no credentials provided for Azure cloud provider
```

AFTER:
```
# /imgpkg pull -i gcr.io/cf-k8s-lifecycle-tooling-klt/accept-imgpkg-b01 -o /tmp/lol --registry-verify-certs=false
Pulling image 'gcr.io/cf-k8s-lifecycle-tooling-klt/accept-imgpkg-b01@sha256:30f38ad891c2a687c4157ff3c1215ed11c0e3be6ae8b48f6a810840f7951220c'
Extracting layer 'sha256:fffbd6f6f0ac9e7fef76b2d21f38a4be513008b24b19ea3992f38661459ad2cb' (1/1)
Succeeded
```

DennisDenuto committed:
refactor:

- rename go file to reflect what its purpose is

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto commented:
@cppforlife 

## Azure manual testing:

I chose to provide client id and secret however the config can also be provided with an assigned managed identity ID

(The flag comes from the library https://github.com/vmware-tanzu/carvel-imgpkg/blob/f7fe9601fcb35e5cbaac88783381137ddef707ab/vendor/github.com/vdemeester/k8s-pkg-credentialprovider/azure/azure_credentials.go#L44)

```
kubectl debug -t -i --image=k8slt/imgpkg-linux:latest node/node1

cat <<EOF >/tmp/azure
{
        "aadClientId": "client",
        "aadClientSecret": "secret"
    }
EOF


/imgpkg pull -i registry-subdomain.azurecr.io/some-image:latest -o /tmp/lol --azure-container-registry-config=/tmp/azure

SUCCEEDS
```


## GCP manual testing:

Created a VM -> with a COS base image

```
/imgpkg pull -i gcr.io/FOO/BAR -o /tmp/lol
E0909 22:38:19.989395    2593 aws_credentials.go:77] while getting AWS credentials NoCredentialProviders: no valid providers in chain. Deprecated.
        For verbose messaging see aws.Config.CredentialsChainVerboseErrors
E0909 22:38:19.997785    2593 gcpcredential.go:74] while reading 'google-dockercfg-url' metadata: http status code: 404 while fetching url http://metadata.google.internal./computeMetadata/v1/instance/attributes/google-dockercfg-url
Pulling image 'gcr.io/cf-k8s-lifecycle-tooling-klt/accept-imgpkg-b01@sha256:30f38ad891c2a687c4157ff3c1215ed11c0e3be6ae8b48f6a810840f7951220c'
Extracting layer 'sha256:fffbd6f6f0ac9e7fef76b2d21f38a4be513008b24b19ea3992f38661459ad2cb' (1/1)
Succeeded
```


### GKE manual testing:

```
kubectl run imgpkg-test -t -i --image=k8slt/imgpkg-linux:latest -- sh -c sleep infinity

root@imgpkg-test:/# /imgpkg pull -i gcr.io/FOO/BAR -o /tmp/lol
E0909 22:47:31.883269    2581 aws_credentials.go:77] while getting AWS credentials NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
E0909 22:47:31.897009    2581 gcpcredential.go:74] while reading 'google-dockercfg-url' metadata: http status code: 404 while fetching url http://metadata.google.internal./computeMetadata/v1/instance/attributes/google-dockercfg-url
Pulling image 'gcr.io/cf-k8s-lifecycle-tooling-klt/accept-imgpkg-b01@sha256:30f38ad891c2a687c4157ff3c1215ed11c0e3be6ae8b48f6a810840f7951220c'
Extracting layer 'sha256:fffbd6f6f0ac9e7fef76b2d21f38a4be513008b24b19ea3992f38661459ad2cb' (1/1)
```



## Feature flag Manual test:

## GCP:
```
IMGPKG_ENABLE_IAAS_AUTH=true ./imgpkg pull -i gcr.io/FOO/BAR -o /tmp/lol
Pulling image 'gcr.io/FOO/BAR@sha256:30f38ad891c2a687c4157ff3c1215ed11c0e3be
6ae8b48f6a810840f7951220c'
Extracting layer 'sha256:fffbd6f6f0ac9e7fef76b2d21f38a4be513008b24b19ea3992f38661459ad2cb' (1/1)
Succeeded
```

```
IMGPKG_ENABLE_IAAS_AUTH=false ./imgpkg pull -i gcr.io/FOO/BAR-o /tmp/lol
imgpkg: Error: Fetching image:
  GET https://gcr.io/v2/FOO/BAR/manifests/latest:
    UNAUTHORIZED:
      You don't have the needed permissions to perform this operation, and you may have invalid credentials. To aut
henticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authenticati
on
```

# AWS Manual testing:

```
root@imgpkg-test:/# /imgpkg pull -i FOO/BAR -o /tmp/lol
Pulling image 'FOO/BAR@sha256:a2d2f6aa77956e4f3053f836dec921f403353032eb8a8927aa9020dfebbec0d4'
Extracting layer 'sha256:cedd63bc9db782e1a2dad92cec779a4a5055b3c026e48e9bb83e7770a267bec4' (1/1)

Succeeded


[cloudshell-user@ip-10-1-13-214 ~]$ imgpkg push -i FOO/BAR -f blah --registry-username AWS --registry-password $PASS
file: blah
Pushed 'FOO/BAR@sha256:a2d2f6aa77956e4f3053f836dec921f403353032eb8a8927aa9020dfebbec0d4'
Succeeded


root@imgpkg-test:/# IMGPKG_ENABLE_IAAS_AUTH=false /imgpkg pull -i FOO/BAR -o /tmp/lol
imgpkg: Error: Fetching image:
  GET https://FOO/v2/BAR/manifests/latest: unexpected status code 401 Unauthorized:
    Not Authorized

```

DennisDenuto committed:
ux: Suppress output from klog

- klog logger is set to 'DiscardLogger'

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto commented:
As part of this PR, logs from iaas providers have been suppressed.

After https://github.com/vmware-tanzu/carvel-imgpkg/pull/241 has been merged, it would be useful to print out IaaS logs *if* the `--debug` flag is set.
